### PR TITLE
lua: Add buffer, window and tab accessors

### PR DESCRIPTION
**Pull request by tjdevries**
@haorenW1025 PR here. Is this what you were thinking?

### CODE REVIEW

**Comment by haorenW1025** [comment]
Yeah exactly! Thanks for doing that:)

**Comment by tjdevries** [comment]
@teto Any reason not to merge this?

**Comment by h-michael** [comment]
note: `vim.{g,v,o}` is still undocumented.
I don't think you need to add it in this PR, but I make a comment so that we do not forget.

**Comment by tjdevries** [comment]
@h-michael Yeah, the problem is that the documentation generator we currently have doesn't know how to do anything besides functions. So if we add it in now, next time someone runs the generator, will get destroyed haha. So I'm working on that in parallel. It's a good thing to add here though, thanks!

File: test/functional/lua/vim_spec.lua [hunk 1]
```diff
@@ -775,6 +775,63 @@ describe('lua stdlib', function()
...
+
+    eq('hi', funcs.luaeval "vim.b.testing")
+    eq(123, funcs.luaeval "vim.b.other")
+    eq(NIL, funcs.luaeval "vim.b.nonexistant")
+
+    exec_lua [[
+    vim.cmd "vnew"
+    ]]
+
+    eq(NIL, funcs.luaeval "vim.b.testing")
```
Comment (jamessan) [inline test/functional/lua/vim_spec.lua:792]: Should the tests also include setting/deleting variables?

Comment (tjdevries) [reply test/functional/lua/vim_spec.lua:792]: Ah, yeah. Why not. I can add it.

**Review by jamessan — APPROVED** [review]

**Comment by justinmk** [comment]
IIRC these were intentionally left out because they are context-dependent. Instead we were aiming for something like `vim.bufs[42].my_var`, `vim.tabs[42].my_var`,  ...

**Comment by tjdevries** [comment]
Hmm, but that's not `b:` works in viml

**Comment by tjdevries** [comment]
Do you want me to add those as well? They could be nice to have (and really easy to add at this point). 

I missed being able to easily use `b:` in Lua though for plugins, so I still am in favor of how this one went (but can remove if you think that it'd be better long term).

**Comment by justinmk** [comment]
> Do you want me to add those as well? They could be nice to have (and really easy to add at this point).

sure, I don't recall any reason why we don't have them. For consistency with API
names I suggest `bufs`, `tabs`, `wins`.

> I missed being able to easily use b: in Lua though for plugins, so I still am
> in favor of how this one went (but can remove if you think that it'd be better
> long term).

we have `vim.bo` already so it's hard to argue against `vim.b`.

**Comment by tjdevries** [comment]
Yes, `vim.bo` was one of the things made me think we could just do this as well (since I looked at some of the impl for that). 

And OK, I will make a PR for those later and ping you.

**Pull request merged**